server/src/main/java/org/candlepin/controller/ActualRefresher.java

```diff
@@ -0,0 +1,1297 @@
+                .map(entry -> entry.getKey())
+                .collect(Collectors.toSet());
+
+            Map versions = this.ownerContentCurator.getVersionedContentById(owner, cids);
+
+            vMap.addEntities(EntityNode.Type.CONTENT,
+                (Map<String, Set<? extends AbstractHibernateObject>>) versions);
+        }
+
+        // Process our nodes for real this timm
+        while (rootIterator.hasNext()) {
+            EntityNode node = rootIterator.next();
+            this.processNode(mapper, vMap, node);
+        }
+
+        // At this point our nodes should be completely processed. One final walk through it to grab the
+        // updated entities should be all we need (as well as DB hits to update ancillary refs)
+
+
+        // This is slow; update the streamNodes method to not exist or return an iterator much like
+        // RootIterator does
+        for (EntityNode node : mapper.streamNodes().collect(Collectors.toList())) {
+            if (node.changed()) {
+                switch (node.getType()) {
+                    case CONTENT:
+                        // TODO: Update this to use the correct curator instead of getting the EM directly
+
+                        this.ownerContentCurator.getEntityManager().merge((Content) node.getUpdatedEntity());
+
+                        if (node.isNewEntity()) {
+                            result.addCreatedContent((Content) node.getUpdatedEntity());
+                        }
+                        else {
+                            result.addUpdatedContent((Content) node.getUpdatedEntity());
+                        }
+
+                        break;
+
+                    case PRODUCT:
+                        this.ownerProductCurator.getEntityManager().merge((Product) node.getUpdatedEntity());
+
+                        if (node.isNewEntity()) {
+                            result.addCreatedProduct((Product) node.getUpdatedEntity());
+                        }
+                        else {
+                            result.addUpdatedProduct((Product) node.getUpdatedEntity());
+                        }
+
+                        break;
+
+                    default:
+                        throw new IllegalStateException("Unexpected node type: " + node.getType());
+                }
+            }
+        }
+
+        this.ownerProductCurator.flush();
+
+        // TODO: Add owner-product and owner-content reference updating/creation
+        // TODO: Update external references
+
+        return result;
+    }
+
+    private EntityNode buildNode(NodeMapper mapper, EntityNode parent, ProductInfo existingEntity) {
+        String entityId = existingEntity.getId();
+        EntityNode<Product, ProductInfo> node = (EntityNode<Product, ProductInfo>) mapper
+            .getNode(EntityNode.Type.PRODUCT, entityId);
+
+        // Check if we already have a node for this entity
+        if (node != null) {
+            return node;
```

Comment:
Correct, if you do call this in the wrong order things break... but don't. :P    Since this is a private method, I would argue that the order is up to callers within the class to get right. 

The whole of buildNode is setup to be recursive and favor existing nodes first and foremost, but then create nodes for net-new products last. That said, it may be correct to ensure both existing product/content entity collections are processed first, rather than existing, imported products and then existing, imported content to avoid the scenario you've outlined above.